File: src/core/inventory.rs
// src/core/inventory.rs
use std::fs;
use std::path::{Path, PathBuf};
use anyhow::Result;
use crate::{defs, conf::config};

#[derive(Debug, Clone)]
pub struct Module {
    pub id: String,
    pub source_path: PathBuf,
    pub mode: String, // "auto", "magic", etc.
}

/// Scans the source directory for enabled modules.
/// Does not access the runtime storage.
pub fn scan(source_dir: &Path, config: &config::Config) -> Result<Vec<Module>> {
    let mut modules = Vec::new();
    if !source_dir.exists() {
        return Ok(modules);
    }

    let module_modes = config::load_module_modes();

    for entry in fs::read_dir(source_dir)? {
        let entry = entry?;
        let path = entry.path();
        
        if !path.is_dir() { continue; }
        
        let id = entry.file_name().to_string_lossy().to_string();
        
        // Skip internal or system directories
        if id == "meta-hybrid" || id == "lost+found" || id == ".git" { continue; }

        // Check for disable flags
        if path.join(defs::DISABLE_FILE_NAME).exists() || 
           path.join(defs::REMOVE_FILE_NAME).exists() || 
           path.join(defs::SKIP_MOUNT_FILE_NAME).exists() { 
            continue; 
        }

        let mode = module_modes.get(&id).cloned().unwrap_or_else(|| "auto".to_string());

        modules.push(Module {
            id,
            source_path: path,
            mode,
        });
    }
    modules.sort_by(|a, b| b.id.cmp(&a.id));

    Ok(modules)
}

File: src/core/inventory.cpp
// core/inventory.cpp - Module inventory implementation
#include "inventory.hpp"
#include "../defs.hpp"
#include "../utils.hpp"
#include <algorithm>
#include <fstream>
#include <sstream>

namespace hymo {

static void parse_module_prop(const fs::path& module_path, Module& module) {
    fs::path prop_file = module_path / "module.prop";
    if (!fs::exists(prop_file)) return;

    std::ifstream file(prop_file);
    std::string line;
    while (std::getline(file, line)) {
        size_t eq = line.find('=');
        if (eq == std::string::npos) continue;

        std::string key = line.substr(0, eq);
        std::string value = line.substr(eq + 1);

        // Trim whitespace if needed (though usually module.prop is strict)
        // Simple key matching
        if (key == "name") module.name = value;
        else if (key == "version") module.version = value;
        else if (key == "author") module.author = value;
        else if (key == "description") module.description = value;
    }
}

std::vector<Module> scan_modules(const fs::path& source_dir, const Config& config) {
    std::vector<Module> modules;
    
    if (!fs::exists(source_dir)) {
        return modules;
    }
    
    try {
        for (const auto& entry : fs::directory_iterator(source_dir)) {
            if (!entry.is_directory()) {
                continue;
            }
            
            std::string id = entry.path().filename().string();
            
            // Skip internal/system directories
            if (id == "hymo" || id == "lost+found" || id == ".git") {
                continue;
            }
            
            // Check for disable flags
            if (fs::exists(entry.path() / DISABLE_FILE_NAME) ||
                fs::exists(entry.path() / REMOVE_FILE_NAME) ||
                fs::exists(entry.path() / SKIP_MOUNT_FILE_NAME)) {
                continue;
            }
            
            // Determine mode
            std::string mode = "auto";
            auto it = config.module_modes.find(id);
            if (it != config.module_modes.end()) {
                mode = it->second;
            }
            
            Module mod{id, entry.path(), mode};
            parse_module_prop(entry.path(), mod);
            modules.push_back(mod);
        }
        
        // Sort by ID descending (Z->A) for overlay priority
        std::sort(modules.begin(), modules.end(), 
            [](const Module& a, const Module& b) {
                return a.id > b.id;
            });
            
    } catch (const std::exception& e) {
        LOG_ERROR("Failed to scan modules: " + std::string(e.what()));
    }
    
    return modules;
}

} // namespace hymo
